func.go

```diff
@@ -292,6 +317,351 @@ func RegisterFunc(fptr interface{}, cfn uintptr) {
+				}
+				switch f.Kind() {
+				case reflect.Struct:
+					place(f)
+				case reflect.Bool:
+					if f.Bool() {
+						val |= 1
+					}
+					shift += 8
+					class |= INTEGER
+				case reflect.Pointer:
+					placedOnStack = true
+					break loop
+				case reflect.Int8:
+					val |= uint64(f.Int()&0xFF) << shift
+					shift += 8
+					class |= INTEGER
+				case reflect.Int16:
+					val |= uint64(f.Int()&0xFFFF) << shift
+					shift += 16
+					class |= INTEGER
+				case reflect.Int32:
+					val |= uint64(f.Int()&0xFFFF_FFFF) << shift
+					shift += 32
+					class |= INTEGER
+				case reflect.Int, reflect.Int64:
+					addInt(uintptr(f.Int()))
+					shift = 0
+					class = NO_CLASS
+				case reflect.Uint8:
+					val |= f.Uint() << shift
+					shift += 8
+					class |= INTEGER
+				case reflect.Uint16:
+					val |= f.Uint() << shift
+					shift += 16
+					class |= INTEGER
+				case reflect.Uint32:
+					val |= f.Uint() << shift
+					shift += 32
+					class |= INTEGER
+				case reflect.Uint, reflect.Uint64:
+					addInt(uintptr(f.Uint()))
+					shift = 0
+					class = NO_CLASS
+				case reflect.Float32:
+					val |= uint64(math.Float32bits(float32(f.Float()))) << shift
+					shift += 32
+					class |= SSE
+				case reflect.Float64:
+					if v.Type().Size() > 16 {
+						placedOnStack = true
+						break loop
+					}
+					addFloat(uintptr(math.Float64bits(f.Float())))
+					class = NO_CLASS
+				case reflect.Array:
+					place(f)
+				default:
+					panic("purego: unsupported kind " + f.Kind().String())
+				}
+			}
+		}
+		place(v)
+		if !flushed {
+			if class == SSE {
+				addFloat(uintptr(val))
+			} else {
+				addInt(uintptr(val))
+			}
+		}
+		if placedOnStack {
+			// reset any values placed in registers
+			*numFloats = savedNumFloats
+			*numInts = savedNumInts
+			*numStack = savedNumStack
+			var placeStack func(v reflect.Value)
+			placeStack = func(v reflect.Value) {
+				for i := 0; i < v.Type().NumField(); i++ {
+					f := v.Field(i)
+					switch f.Kind() {
+					case reflect.Pointer:
+						addStack(f.Pointer())
+					case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+						addStack(uintptr(f.Int()))
+					case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+						addStack(uintptr(f.Uint()))
+					case reflect.Float32:
+						addStack(uintptr(math.Float32bits(float32(f.Float()))))
+					case reflect.Float64:
+						addStack(uintptr(math.Float64bits(f.Float())))
+					case reflect.Struct:
+						placeStack(f)
+					default:
+						panic("purego: unsupported kind " + f.Kind().String())
+					}
+				}
+			}
+			placeStack(v)
+		}
+		return keepAlive
+	}
+	panic("purego: struct has field that can't be allocated")
+}
+
+func roundUpTo8(val uintptr) uintptr {
+	return (val + 7) &^ 7
+}
+
+// isHVA reports a Homogeneous Floating-point Aggregate (HFA) which is a Fundamental Data Type that is a
+// Floating-Point type and at most four uniquely addressable members (5.9.5.1).
```

Comment:
What is `5.9.5.1`?

Reply:
It the section in the spec where information about the type is

Reply:
Let's add a link to the spec. Only from this comment, it is hard which PDF is referred.